doc/i2c_bus_lock_prevention/i2c_bus_lock_prevention_hld.md

```diff
@@ -0,0 +1,425 @@
+
+Field: "device_name"
+
+Refer to [6.2. I2C Isolated Region](#62-i2c-isolated-region) for the consistent of the name in "I2C Isolated Region ID".
+"Isolated I2C Device Address" is the address of the isolated i2c device. The address is got from the platform API. The platform
+API may output the corresponding i2c device address under '/sys/bus/i2c/devices/'.
+
+Here is an example for the record in "I2C_ISOLATION_LIST" table in STATE DB.
+The table below contains one faulty device. The I2C Isolated Region ID of the device is "77-2-77-6"
+The isolated i2c device address in the region is "30-0050".
+The device name of the isolated i2c device is "Ethernet40". That means the transceiver inserted in
+the port Ethernet40 is being isolated.
+```
+admin@sonic:~$ redis-cli -n 6 hgetall "I2C_ISOLATION_LIST|77-2-72-6|30-0050"
+1) "device_name"
+2) "Ethernet40"
+```
+
+### 8. SAI API 
+
+SAI API is not used in this feature.
+
+### 9. Configuration and management 
+
+No configuration in this feature. 
+
+### 10. Warmboot and Fastboot Design Impact  
+
+The feature does not support warmboot and fastboot, and it has nothing to do with warmboot and fastboot.
+
+### 11. Memory Consumption
+
+No memory consumption is expected when the feature is disabled via compilation and no growing memory consumption while feature is disabled by configuration.
+
+### 12. Restrictions/Limitations  
+
+The I2C bus lock prevention only works when the swappable faulty device is located behind a I2C Mux's channel.
+
+#### 12.1. Minimum Interval to Detect the Presence State Change of Swapable Modules like Transceivers, PSUs and FAN Modules
+
+Due to the implemenation of xcvrd, psud and thermalctld, they polls for the presence state at fixed interval. If an inserted module is removed and replaces with a new module in a short time, the daemon might not be able to realize the presence state has ever been changed to not present because it does not poll the presence state when the module is removed.
+
+The mechanism for removing devices from isolation depends on the presence state change. Therefore, for the I2C health daemon to detect the presence state change, it must remove the module and wait for the detection interval to elapse to ensure that the "not present" state can be detected.
+
+### 13. Testing Requirements/Design  
+
+#### 13.1. Unit Test cases  
+
+#### 13.2. System Test cases
+
+Test Case 1: The i2c healthd shall only runs on the platform that support i2c bus lock prevention.
+
+Test Case 2: Plug a faulty transceiver on the device. The i2c healthd shall detect the faulty transceiver, isolate the device and recover the i2c bus.
+
+Test Case 3: Unplug a faulty transceiver. The faulty transceiver shall be removed from the isolation list.
+
+Test Case 4: Power off the device. Plug a faulty transceiver on the device and power on the device. The i2c healthd shall detect the faulty transceiver, isolate the device and recover the i2c bus.
+

```

Comment:
Additional test cases:
- Simulate a faulty temperature sensor. Ensure that fans go to max RPM. 
- Recover the faulty temperature sensor. Ensure that fans resume normal operation.
- Simulate other types of faulty devices, like fans, fan trays, and PSUs. 
- Warmboot/fastboot cases if applicable (see my earlier comment).
- Swap multiple times between a faulty and good transceiver, and ensure that `i2chealthd` operates as designed. 
- Simulate a working but busy i2c bus. For example, a fan whose target RPM is constantly being updated due to environmental temperature fluctuations. Will `i2chealthd` consider it locked up?